Automattic/woocommerce-payments
# New Tracking events and passing more tracking parameters

Closes https://github.com/Automattic/woocommerce-payments-server/issues/4535

#### Changes proposed in this Pull Request

* Add new type defs for `wcSettings`.
* Add a note about the default event properties added by Woo, and avoid us overwriting them.
* Update both PHP and JS record event functions to add extra default details – `jetpack_connected`, `wcpay_version`, and `woo_country_code`.
* Add new events:
  * `wcadmin_wcpay_account_connect_wpcom_connection_start` – for the start of the Jetpack redirect.
  * `wcadmin_wcpay_account_connect_wpcom_connection_failure` – for a failed Jetpack connection attempt.
  * `wcadmin_wcpay_gateway_toggle` – for WooPayments enable and disable in WooPayments settings and WooCommerce payments settings.

Ensure all client events in this doc: https://docs.google.com/spreadsheets/d/1a-cRDpwWsZXYLpe7gxUxG2jduFvTFuXp-wCdM14eqKE/edit#gid=0 are accounted for.

#### Testing instructions

##### Prerequisites

* Run `npm run watch`.
* Install the latest version of WooCommerce (+8.4) on your store using the plugin update page.
* Enable tracking in **WooCommerce -> Settings -> Advanced -> Woo.com**

Instructions borrowed from https://github.com/Automattic/woocommerce-payments/pull/6128:

> To check that Tracks events are properly registered, you can use one of the following methods:
> 
> Run localStorage.setItem( 'debug', 'wc-admin:tracks' ); on your site browser console. And look for verbose logs.
> Install [Tracks Vigilante](https://github.com/Automattic/tracks-chrome-extension) browser extension and use its log of events.
> Go to [Tracks Live view](https://mc.a8c.com/tracks/live/), filter by any of the allowed methods, and wait patiently (~10min) for events to be registered there.
> Go through the onboarding process ensuring every new track event gets registered accordingly. Apart from the description you can see [a table with screenshots here](https://docs.google.com/document/d/1XXPpXJtYPKkmOQxD43el8Mk6JTuw1IGG1SskKOvgYHE/edit).

To test this PR, using the instructions above, go through an onboarding, and ensure the following params are tracked in all onboarding events:

- `is_test_mode`
- `jetpack_connected`
- `wcpay_version`
- `woo_country_code`

As a bonus, you can also test the following:

* Use the JetPack Debug Tools to break the JP connection to your store.
* Start a new onboarding.
* When you get to Jetpack, click to return back to the store without creating a connection.
* Verify that the `wcadmin_wcpay_account_connect_wpcom_connection_failure ` event was tracked.
* Start another one, verify the `wcpay_account_connect_wpcom_connection_start` event gets tracked at the start of the Jetpack connection.
* Verify the `wcpay_account_connect_wpcom_connection_success` event gets tracked at the end.

-------------------


**Post merge**
